src/logic/map.cc

```diff
@@ -540,154 +540,291 @@ void Map::set_origin(const Coords& new_origin) {
-					imm->remove(egbase);
-				}
-				while (Bob* bob = field.get_first_bob()) {
-					bob->remove(egbase);
-				}
-			}
-			if (x < w && y < h) {
-				if (Coords c_old = transform_coords(c_new, split, width_, height_, w, h)) {
-					bool& entry = preserved_coords[get_index(c_old, width_)];
-					if (!entry) {
-						// Copy existing field
-						entry = true;
-						new_fields[get_index(c_new, w)] = operator[](c_old);
-						continue;
-					}
-				}
-				// Init new field
-				Field& field = new_fields[get_index(c_new, w)];
+	for (int16_t x = 0; x < w; ++x) {
+		for (int16_t y = 0; y < h; ++y) {
+			Coords cn(x, y);
+			if (Coords co = lambda_co(x, y)) {
+				new_fields[get_index(cn, w)] = operator[](co);
+				assert(!was_preserved[get_index(co)]);
+				was_preserved[get_index(co)] = true;
+			} else {
+				Field& field = new_fields[get_index(cn, w)];
 				field.set_height(10);
 				field.set_terrains(default_terrains);
 			}
 		}
 	}
+	for (Coords& c : starting_pos_) {
+		if (c) {  // only if set (!= Coords::null())
+			c = lambda_cn(c);
+		}
+	}
 
-	// Replace all fields
-	fields_.reset(new Field[field_size]);
-	clear_array<>(&fields_, field_size);
-	for (size_t ind = 0; ind < field_size; ++ind) {
-		fields_[ind] = new_fields[ind];
+	{
+		PortSpacesSet new_port_spaces;
+		for (const Coords& c : port_spaces_) {
+			if (Coords cn = lambda_cn(c)) {
+				new_port_spaces.insert(cn);
+			}
+		}
+		port_spaces_ = new_port_spaces;
+	}
+
+	for (int16_t x = 0; x < width_; ++x) {
+		for (int16_t y = 0; y < height_; ++y) {
+			Coords c(x, y);
+			if (!was_preserved[get_index(c)]) {
+				Field& f = operator[](c);
+				if (upcast(Immovable, i, f.get_immovable())) {
+					i->remove(egbase);
+				}
+				while (Bob* b = f.get_first_bob()) {
+					b->remove(egbase);
+				}
+			}
+		}
 	}
-	log("Resized map from (%d, %d) to (%u, %u) at (%d, %d)\n", width_, height_, w, h, split.x,
-	    split.y);
+
 	width_ = w;
 	height_ = h;
+	fields_.reset(new_fields.release());
+	// Always call allocate_player_maps() while changing the map's size. Forgetting to do so will
+	// result in random crashes.
+	egbase.allocate_player_maps();
+	recalc_whole_map(egbase);
 
-	// Inform immovables and bobs about their new position
-	for (MapIndex idx = 0; idx < field_size; ++idx) {
-		Field& f = operator[](idx);
-		if (upcast(Immovable, immovable, f.get_immovable())) {
-			immovable->position_ = get_fcoords(f);
-		}
-		// Ensuring that all bob iterators are changed correctly is a bit hacky, but the more obvious
-		// solution of doing it like in set_origin() is highly problematic here, or so ASan tells me
-		std::vector<Bob*> bobs;
-		for (Bob* bob = f.get_first_bob(); bob; bob = bob->get_next_bob()) {
-			bobs.push_back(bob);
-		}
-		f.bobs = nullptr;
-		for (Bob* bob : bobs) {
-			bob->position_.field = nullptr;
-			bob->linknext_ = nullptr;
-			bob->linkpprev_ = nullptr;
-			bob->set_position(egbase, get_fcoords(f));
+	for (int16_t x = 0; x < width_; ++x) {
```

Comment:
What kind of cleanup happens here? e.g.
// Recalculate FCords for Immovables (Stones ...) and BOBs (Animals, trees ...) as backlink may got lost